src/Microsoft.ML.Tokenizers/Model/Tiktoken.cs

```diff
@@ -81,48 +73,60 @@ public Tiktoken(Stream tikTokenBpeFileStream, IReadOnlyDictionary<string, int>?
-        }
-
-        private void Initialize(Stream tikTokenBpeFileStream, IReadOnlyDictionary<string, int>? specialTokensEncoder = null)
+        private Tiktoken(Stream tikTokenBpeFileStream, IReadOnlyDictionary<string, int>? specialTokensEncoder, int cacheSize, bool disposeStream) : this(cacheSize)
         {
-            if (tikTokenBpeFileStream is null)
+            try
             {
-                throw new ArgumentNullException(nameof(tikTokenBpeFileStream));
-            }
-
-            (_encoder, _vocab, _decoder) = LoadTikTokenBpe(tikTokenBpeFileStream);
+                (_encoder, _vocab, _decoder) = LoadTikTokenBpeAsync(tikTokenBpeFileStream, useAsync: false).GetAwaiter().GetResult();
 
-            _specialTokensEncoder = specialTokensEncoder;
-            if (_specialTokensEncoder is not null)
+                _specialTokensEncoder = specialTokensEncoder;
+                if (_specialTokensEncoder is not null)
+                {
+                    _specialTokensDecoder = _specialTokensEncoder.ToDictionary(kvp => kvp.Value, kvp => kvp.Key);
+                }
+            }
+            finally
             {
-                _specialTokensDecoder = _specialTokensEncoder.ToDictionary(kvp => kvp.Value, kvp => kvp.Key);
+                if (disposeStream)
+                {
+                    tikTokenBpeFileStream.Dispose();
+                }
             }
         }
 
+        private Tiktoken(int cacheSize)
+        {
+            _cache = new LruCache<string, int[]>(cacheSize);
+        }
+
         /// <summary>
         /// Load BPE rank dictionary from a stream.
         /// </summary>
         /// <param name="tikTokenBpeFileStream">Stream to the BPE rank file</param>
+        /// <param name="useAsync">Whether to perform I/O synchronously or asynchronously.</param>
         /// <returns>Map of byte[] to integer token id</returns>
         /// <exception cref="InvalidOperationException"></exception>
-        internal static (Dictionary<byte[], int>, Dictionary<string, int>, IReadOnlyDictionary<int, byte[]>) LoadTikTokenBpe(Stream tikTokenBpeFileStream)
+        internal static async ValueTask<(Dictionary<ReadOnlyMemory<byte>, int>, Dictionary<string, int>, IReadOnlyDictionary<int, byte[]>)> LoadTikTokenBpeAsync(Stream tikTokenBpeFileStream, bool useAsync)
         {
-            var encoder = new Dictionary<byte[], int>(new ByteArrayComparer());
+            var encoder = new Dictionary<ReadOnlyMemory<byte>, int>(ReadOnlyMemoryByteComparer.Instance);
             var vocab = new Dictionary<string, int>();
             var decoder = new Dictionary<int, byte[]>();
 
             try
             {
                 using (StreamReader reader = new StreamReader(tikTokenBpeFileStream))
                 {
-                    while (!reader.EndOfStream)
+                    while (true)
                     {
-                        string? line = reader.ReadLine();
+                        string? line = useAsync ?
```

Comment:
This is more of a reliability/scalability fix rather than throughput. The asynchronous creation code path and the synchronous creation code path were both sharing this routine, which resulted in the asynchronous code path doing synchronous I/O (on a network stream). Now the async path does async and the sync path does sync, still sharing this same routine.

Reply:
Why we don't always do async operation here regardless if the code is coming from sync or async callers?

Reply:
Because if this is async, a sync caller will be forced to block this thread into the operation completes. The operation could easily require a thread pool thread to complete, yet if this is on a thread pool thread, it'd be blocking one of the very resources needed to make forward progress. The thing less scalable than sync I/O is sync-over-async I/O.

Reply:
But I thought you wanted to have a constructor which execute async too. no?

Reply:
I do (a factory method rather than a ctor, but yeah). It will use this.